Source/santa-driver/SantaPrefixTree.cc

```diff
@@ -14,26 +14,61 @@
 
 #include "SantaPrefixTree.h"
 
+#ifdef KERNEL
+#include <libkern/locks.h>
+
 #include "SNTLogging.h"
+#else
+#define lck_grp_attr_alloc_init() nullptr
+#define lck_grp_alloc_init(name, attr) nullptr
+#define lck_attr_alloc_init() nullptr
+
+#define lck_attr_free(attr) do {} while(0)
+#define lck_grp_free(grp) do {} while(0)
+#define lck_grp_attr_free(grp_attr) do {} while(0)
+
+#define lck_rw_lock_shared(l) l.lock_shared()
+#define lck_rw_unlock_shared(l) l.unlock_shared()
+#define lck_rw_lock_exclusive(l) l.lock()
+#define lck_rw_unlock_exclusive(l) l.unlock()
 
-SantaPrefixTree::SantaPrefixTree() {
+#define lck_rw_lock_shared_to_exclusive(l) [this] () { l.unlock_shared(); return false;}()
+#define lck_rw_lock_exclusive_to_shared(l) l.unlock(); l.lock_shared()
+
+#define lck_mtx_lock(l) l.lock()
+#define lck_mtx_unlock(l) l.unlock()
+#endif // KERNEL
+
+SantaPrefixTree::SantaPrefixTree(uint32_t max_nodes) {
   root_ = new SantaPrefixNode();
   node_count_ = 0;
+  max_nodes_ = max_nodes;
 
   spt_lock_grp_attr_ = lck_grp_attr_alloc_init();
   spt_lock_grp_ = lck_grp_alloc_init("santa-prefix-tree-lock", spt_lock_grp_attr_);
   spt_lock_attr_ = lck_attr_alloc_init();
+
+  #ifdef KERNEL
```

Comment:
Could you avoid these ifdef's by defining lck_rw_alloc_init, lck_rw_t, lck_mtx_t, etc.?

```cpp
#define lck_rw_t std::shared_mutex
#define lck_mtx_t std::mutex

#define lck_rw_alloc_init(a, b) new std::shared_mutex
#define lck_rw_lock_shared(l) l->lock_shared()
...

#define lck_mtx_alloc_init(a, b) new std::mutex
#define lck_mtx_lock(l) l->lock();
```

Reply:
The templates get angry: `No viable overloaded '='`